backend/src/main/scala/cromwell/backend/package.scala

```diff
@@ -61,5 +64,44 @@ package object backend {
 
   case class PreemptedException(msg: String) extends Exception(msg)
 
+  /**
+    * @param usedInCallCaching Whether or not this attribute is used to determine a cache hit or miss. ALL attributes are hashed and stored.
+    */
   case class RuntimeAttributeDefinition(name: String, required: Boolean, default: Option[WdlValue], usedInCallCaching: Boolean)
+
+  object RuntimeAttributeDefinition {
+
+    def evaluateRuntimeAttributes(unevaluated: RuntimeAttributes,
+                                  wdlFunctions: WdlStandardLibraryFunctions,
+                                  evaluatedInputs: Map[LocallyQualifiedName, WdlValue]): Try[Map[String, WdlValue]] = {
+      val tryInputs = evaluatedInputs map { case (x, y) => x -> Success(y) }
+      val mapOfTries = unevaluated.attrs mapValues {
+        expr => expr.evaluate(buildMapBasedLookup(tryInputs), wdlFunctions)
+      }
+      TryUtil.sequenceMap(mapOfTries)
+    }
+
+    def buildMapBasedLookup(evaluatedDeclarations: Map[LocallyQualifiedName, Try[WdlValue]])(identifier: String): WdlValue = {
+      val successfulEvaluations = evaluatedDeclarations collect {
+        case (k, v) if v.isSuccess => k -> v.get
+      }
+      successfulEvaluations.getOrElse(identifier, throw new WdlExpressionException(s"Could not resolve variable $identifier as a task input"))
+    }
+
+    def addDefaultsToAttributes(runtimeAttributeDefinitions: Set[RuntimeAttributeDefinition], workflowOptions: WorkflowOptions)
+                               (specifiedAttributes: Map[LocallyQualifiedName, WdlValue]): Map[LocallyQualifiedName, WdlValue] = {
+      import WdlValueJsonFormatter._
+      import spray.json._   // IGNORE INTELLIJ - this *is* required (unless it isn't any more, who will ever know...!)
+
+      def isUnspecifiedAttribute(name: String) = !specifiedAttributes.contains(name)
+
+      val missing = runtimeAttributeDefinitions filter { x => isUnspecifiedAttribute(x.name) }
+      val defaults = missing map { x => (x, workflowOptions.getDefaultRuntimeOption(x.name)) } collect {
+        case (runtimeAttributeDefinition, Success(jsValue)) => runtimeAttributeDefinition.name -> jsValue.convertTo[WdlValue]
+        case (RuntimeAttributeDefinition(name, _, Some(defaultValue), _), _) => name -> defaultValue
```

Comment:
The use of `defaultValue` terminology for compiled-in defaults as opposed to the per-workflow default runtime attributes is kind of confusing